src/Microsoft.Extensions.Configuration.AzureAppConfiguration/SrvLookupClient.cs

```diff
@@ -0,0 +1,145 @@
+    {
+        private class OriginHostCacheItem
+        {
+            public string OriginHost { get; set; }
+
+            public DateTimeOffset CacheExpires { get; set; }
+        }
+
+        private readonly ConcurrentDictionary<string, OriginHostCacheItem> _cachedOriginHosts;
+
+        private readonly LookupClient _tcpLookupClient;
+        private readonly LookupClient _udpLookupClient;
+
+        const string TcpOrigin = "_origin._tcp";
+        const string TCP = "_tcp";
+        const string Alt = "_alt";
+        const int MaxSrvRecordCountPerRecordSet = 20;
+
+        private readonly Logger _logger;
+
+        public SrvLookupClient(Logger logger)
+        {
+            _cachedOriginHosts = new ConcurrentDictionary<string, OriginHostCacheItem>();
+
+            _udpLookupClient = new LookupClient(new LookupClientOptions()
+            {
+                UseTcpFallback = false
+            });
+
+            _tcpLookupClient = new LookupClient(new LookupClientOptions()
+            {
+                UseTcpOnly = true
+            });
+
+            _logger = logger;
+        }
+
+        public async Task<IEnumerable<SrvRecord>> QueryAsync(string host, CancellationToken cancellationToken)
+        {
+            IEnumerable<SrvRecord> resultRecords = Enumerable.Empty<SrvRecord>();
+            var originSrvDns = $"{TcpOrigin}.{host}";
+            
+            bool exists = _cachedOriginHosts.TryGetValue("originSrvDns", out OriginHostCacheItem originHost);
+
+            try
+            {
+                if (!exists || originHost.CacheExpires <= DateTimeOffset.UtcNow)
+                {
+                    var records = await InternalQueryAsync(originSrvDns, cancellationToken).ConfigureAwait(false);
+
+                    if (records == null || records.Count() == 0)
+                    {
+                        return Enumerable.Empty<SrvRecord>();
+                    }
+
+                    if (!exists)
+                    {
+                        originHost = new OriginHostCacheItem()
+                        {
+                            OriginHost = records.First().Target.Value.Trim('.'),
+                            CacheExpires = DateTimeOffset.UtcNow.AddMinutes(30)
+                        };
+
+                        _cachedOriginHosts[originSrvDns] = originHost;
+                    }
+                    else
+                    {
+                        originHost.OriginHost = records.First().Target.Value.Trim('.');
+                        originHost.CacheExpires = DateTimeOffset.UtcNow.AddMinutes(30);
+                    }
+                }
+
+                int index = 0;
+
+                while (true)
+                {
+                    string altSrvDns = $"{TcpAlternative(index)}.{originHost.OriginHost}";
+
+                    var records = await InternalQueryAsync(altSrvDns, cancellationToken).ConfigureAwait(false);
+                    resultRecords = resultRecords.Concat(records);

```

Comment:
This will throw if `records` is null

Reply:
`records` returned from `InternalQueryAsync()` is not supposed to be null.  But it still deserves a null check here.